### PR TITLE
Show current filename in sticky pull request header

### PR DESCRIPTION
**Pull request by DrewML**
Ref: Issue #98 

Todo:
- [x] Unbind scroll/resize listeners on partial page nav to non-PR page
- [x] Only update file name label in DOM when it's changed from previous val
- [x] Hide filename in header when not sticky (top of page)
- [x] Enable on Commit pages (currently only PR pages)
- [x] Determine desired behavior [when sticky header will never reach last commit ](https://github.com/sindresorhus/refined-github/issues/98#issuecomment-200637951)

### CODE REVIEW

**Comment by hkdobrev** [comment]
@DrewML You could use GFM todo lists in the PR description https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments :wink:

**Comment by DrewML** [comment]
@hkdobrev Good call! Should be knocking out the top 3 in the list when I'm home from work tonight.

Still need some help determining what is the most appropriate way to handle the 4th item in that list.

**Comment by hkdobrev** [comment]
> Determine desired behavior [when sticky header will never reach last commit](https://github.com/sindresorhus/refined-github/issues/98#issuecomment-200637951)

According to other libraries and websites that do this kind of things, you should only change the file name in the header only when the contents of the previous one has gone above the viewport. I think the eyeline library for Discourse.org used to do it like that, but that was years ago.

File: extension/content.js [hunk 1]
```diff
@@ -328,6 +368,11 @@ document.addEventListener('DOMContentLoaded', () => {
 			if (isCommit()) {
 				addPatchDiffLinks();
 			}
+
+			if (isPRFiles()) {
```
Comment (hkdobrev) [inline extension/content.js]: @DrewML Ultimately you'd use a check which will catch either `/pull/<id>/files` or `<pull/<id>/commits/<sha>`.

Comment (hkdobrev) [inline extension/content.js]: You could break the `isPR()` and `isCommit()` checks above into smaller pieces using a new one like `isPRCommit()` and then you could combine checks.

Comment (DrewML) [inline extension/content.js]: That's funny - I originally had the POC checking for both, but I had *thought* late last night that the sticky bar wasn't enabled on commit pages. I just double-checked, and I was wrong :)

Comment (hkdobrev) [inline extension/content.js]: It's enabled on:
- PR files listing
- PR commit pages (new)

It's not enabled on:
- commit pages (default)
- compare views

**Comment by DrewML** [comment]
All todo items have been completed/pushed. Let me know if anything needs to change, and let me know if you want me to squash my commits down.

**Comment by paulmolluzzo** [comment]
This is related to lines [343-350](
https://github.com/DrewML/refined-github/blob/b64d0da24249ec98d5a69eda27423c0adbc54e9b/extension/content.js#L343-350), [373](https://github.com/DrewML/refined-github/blob/b64d0da24249ec98d5a69eda27423c0adbc54e9b/extension/content.js#L373), and [401-405](https://github.com/DrewML/refined-github/blob/b64d0da24249ec98d5a69eda27423c0adbc54e9b/extension/content.js#L401-405):

Returning the remove event listener in the function `bindDiffHeaderListeners()` and then going through the `let unbindDiffHeaderListeners;` and `if {} else if {}` logic seems kind of roundabout. It could just be personal preference, but would something like this be a little more to the point?

```js
const headerEvents = ['scroll', 'resize'];
.
.
.
function bindDiffHeaderListeners() {
    headerEvents.forEach(event => window.addEventListener(event, diffHeaderFilename));
}

function unbindDiffHeaderListeners() {
    headerEvents.forEach(event => window.removeEventListener(event, diffHeaderFilename));
};
.
.
.
unbindDiffHeaderListeners();

if (isPRFiles() || isPRCommit()) {
    bindDiffHeaderListeners();
}
```

**Comment by DrewML** [comment]
@paulmolluzzo yeah, that's much simpler, and avoids the awkward state in the `DOMContentLoaded` callback. Will update today

**Comment by hkdobrev** [comment]
I've noticed a buggy behaviour. The first file name is not shown in the header initially, but only after the second file name is shown and then you scroll back up to the first file. If you scroll to the top of the page then and then scroll down the first file name is correctly shown from the initial stuck of the header.

Here's a GIF showing the buggy behaviour (sorry for the broken colours, it's from the recording):

scroll-filename

**Comment by DrewML** [comment]
@hkdobrev Nice catch - I noticed that earlier today when surfing around. I'm pretty sure I know what's causing it - will work on cleaning that up either tonight or tomorrow.

**Comment by DrewML** [comment]
@hkdobrev Fix has been pushed for your reported bug. Let me know if you see any other weirdness.
@paulmolluzzo Switched to bind/unbind functions as suggested.

I don't feel super great about the `diffFile` IIFE, but the only immediate alternatives I could think of were:

1. Checking the DOM each time for the last value (less than ideal)
2. Keeping the state in a `let` assignment in the outermost scope in `content.js`, which _feels_ more hacky than my current solution.

File: extension/content.js [hunk 2]
```diff
@@ -282,6 +284,78 @@ $(document).on('click', event => {
...
+	const hasChanged = nextFile => {
+		if (nextFile !== lastFile) {
+			lastFile = nextFile;
+			return true;
+		}
+
+		return false;
+	};
+
+	const reset = function reset() {
```
Comment (sindresorhus) [inline extension/content.js]: No need to name the function. It's inferred from the variable name.

**Comment by sindresorhus** [comment]
This introduces a lot of global functions. Maybe we should move all the functions to a separate file inside an IIFE and only expose the two "publicly" used functions under a namespace?

```js
diffFileHeader.setup();
diffFileHeader.destroy();
```

File: extension/content.js [hunk 3]
```diff
@@ -7,8 +7,9 @@ const isDashboard = () => location.pathname === '/' || /(^\/(dashboard))/.test(l
 const isRepo = () => /^\/[^/]+\/[^/]+/.test(location.pathname);
 const isRepoRoot = () => location.pathname.replace(/\/$/, '') === `/${ownerName}/${repoName}` || /(\/tree\/)(\w|\d|\.)+(\/$|$)/.test(location.href);
 const isPR = () => /^\/[^/]+\/[^/]+\/pull\/\d+/.test(location.pathname) || /^\/[^/]+\/[^/]+\/pull\/\d+\/commits\/[0-9a-f]{5,40}/.test(location.pathname);
+const isPRCommit = () => /^\/[^/]+\/[^/]+\/pull\/\d+\/commits\/[0-9a-f]{5,40}/.test(location.pathname);
 const isCommit = () => {
```
Comment (sindresorhus) [inline extension/content.js]: @hkdobrev @paulmolluzzo Unrelated to this PR, but I think our paths are getting out of control. How about we make a util.js file that exposes these utilities?

```js
util.isRepo();
util.isRepoRoot();
...
```

File: extension/content.js [hunk 4]
```diff
@@ -282,6 +284,78 @@ $(document).on('click', event => {
...
+		return;
+	}
+
+	const filename = $(targetDiffFile).find('.file-header').attr('data-path');
+
+	if (!diffFile.hasChanged(filename)) {
+		return;
+	}
+
+	$('.diff-toolbar-filename').text(`Filename: ${filename}`);
```
Comment (sindresorhus) [inline extension/content.js]: Is there really any need to prefix it with `Filename: `? That's pretty obvious to me at least.

**Comment by sindresorhus** [comment]
This looks very good and seems to be working perfectly for me 👌

File: extension/content.js (hunk 3 above)
Comment (hkdobrev) [inline extension/content.js]: I'm OK with this as long as we don't use the name `utils`, but something more explicit and with one purpose only like `urlHelper`, `urlUtils` or just `url`.

**Comment by hkdobrev** [comment]
> This introduces a lot of global functions. Maybe we should move all the functions to a separate file inside an IIFE and only expose the two "publicly" used functions under a namespace?

:+1:

File: extension/content.js (hunk 3 above)
Comment (paulmolluzzo) [inline extension/content.js]: :+1: it'll clean up the code a lot.

Comment (hkdobrev) [inline extension/content.js]: We can do that after this PR is merged :)

File: extension/content.js (hunk 4 above)
Comment (DrewML) [inline extension/content.js]: Nope - I can remove the label

**Comment by DrewML** [comment]
> This introduces a lot of global functions. Maybe we should move all the functions to a separate file inside an IIFE and only expose the two "publicly" used functions under a namespace?

Sounds good to me - was starting to feel crowded. Will push an update today.

File: extension/content.js (hunk 3 above)
Comment (sindresorhus) [inline extension/content.js]: > We can do that after this PR is merged :)

:+1: Yes, meant after.

**Comment by DrewML** [comment]
Updates pushed to branch - moved all functionality relating to this PR into a new content script.

**Comment by sindresorhus** [comment]
@DrewML Can you use arrow functions instead of functions and fix the merge conflict? Looks good to me after that :)

**Comment by DrewML** [comment]
@sindresorhus Done! Also squashed the mile of commits down to 1 for the feature.

**Comment by sindresorhus** [comment]
@DrewML This is amazing. Superb work! 

celebration

**Comment by sindresorhus** [comment]
@DrewML Would you be interested in joining this project as a maintainer? No worries if not though :)

**Comment by DrewML** [comment]
@sindresorhus For sure! I'm in.

**Comment by sindresorhus** [comment]
Yay! :)

// @hkdobrev @paulmolluzzo

**Pull request merged**